internal/daemon/vendor-specific-plugins/marvell/main.go

```diff
@@ -33,19 +34,172 @@ const (
+	nfInterfaceName string
+	dpInterfaceName string
+	dpMAC           string
+	health          string
+}
 type mrvlVspServer struct {
 	pb.UnimplementedLifeCycleServiceServer
 	pb.UnimplementedNetworkFunctionServiceServer
 	pb.UnimplementedDeviceServiceServer
 	opi.UnimplementedBridgePortServiceServer
-	log         logr.Logger
-	grpcServer  *grpc.Server
-	wg          sync.WaitGroup
-	done        chan error
-	startedWg   sync.WaitGroup
-	pathManager utils.PathManager
-	version     string
-	isDPUMode   bool
+	log           logr.Logger
+	grpcServer    *grpc.Server
+	wg            sync.WaitGroup
+	done          chan error
+	startedWg     sync.WaitGroup
+	pathManager   utils.PathManager
+	version       string
+	isDPUMode     bool
+	deviceStore   map[string]mrvlDeviceInfo
+	portType      string
+	noOfPortPairs int
+}
+
+// readConfig function to read the config file for the given device
+func readConfig() (deviceConfig, error) {
+	config_path := "/config.yaml"
+	data, err := os.ReadFile(config_path)
+	if err != nil {
+		return deviceConfig{}, err
+	}
+	var config deviceConfig
+	err = yaml.Unmarshal(data, &config)
+	if err != nil {
+		return deviceConfig{}, err
+	}
+	return config, nil
+}
+
+// createVethPair function to create a veth pair with the given index and InterfaceInfo
+func (vsp *mrvlVspServer) createVethPair(index int) error {
+	nfInterfaceName := fmt.Sprintf("nf_interface%d", index)
+	dpInterfaceName := fmt.Sprintf("dp_interface%d", index)
+	vethLink := &netlink.Veth{
+		LinkAttrs: netlink.LinkAttrs{Name: nfInterfaceName},
+		PeerName:  dpInterfaceName,
+	}
+	if err := netlink.LinkAdd(vethLink); err != nil {
+		return err
+	}
+	if err := netlink.LinkSetUp(vethLink); err != nil {
+		return err
+	}
+
+	nfLink, err := netlink.LinkByName(nfInterfaceName)
+	if err != nil {
+		return err
+	}
+
+	peerLink, err := netlink.LinkByName(dpInterfaceName)
+	if err != nil {
+		return err
+	}
+
+	if err := netlink.LinkSetUp(peerLink); err != nil {
+		return err
+	}
+
+	vsp.deviceStore[nfLink.Attrs().HardwareAddr.String()] = mrvlDeviceInfo{
+		nfInterfaceName: nfInterfaceName,
+		dpInterfaceName: dpInterfaceName,
+		dpMAC:           peerLink.Attrs().HardwareAddr.String(),
+		health:          "Healthy",
+	}
+	return nil
+}
+
+func (vsp *mrvlVspServer) createHwLBK() error {
+	//TODO: Implement HW Loopback
+	klog.Infof("currently only veth pairs are supported")
+	return errors.New("currently only veth pairs are supported")
+}
+
+// CleanVethPairs function to clean all the veth pairs created
+func (vsp *mrvlVspServer) CleanVethPairs() error {
+	if vsp.deviceStore != nil {
+		for nfMacAddress, mrvlDeviceInfo := range vsp.deviceStore {
+			nfLink, err := netlink.LinkByName(mrvlDeviceInfo.nfInterfaceName)
+			if err != nil {
+				klog.Errorf("Error occurred in getting Link By Name: %v", err)
+				return err
+			}
+			if err := netlink.LinkDel(nfLink); err != nil {
+				klog.Errorf("Error occurred in deleting Link: %v", err)
+				return err
```

Comment:
Shouldn't the cleanup proceed trying to delete all interfaces, even if errors happen? Aborting on the first error seems wrong, especially since iterating over `range vs.deviceStore` iterates in an arbitrary order.

Also, I think the map must not be modified (`deletet()`) while iterating over it.

I think the errors should be all collected, but the code should also try to delete all interfaces:
```go
    var errResult error
    if vsp.deviceStore != nil {
        deviceStore := vsp.deviceStore
        vsp.deviceStore = nil
        for nfMacAddress, mrvlDeviceInfo := range deviceStore {
             nfLink, err := netlink.LinkByName(mrvlDeviceInfo.nfInterfaceName)
             if err != nil {
                 klog.Errorf("Error occurred in deleting Link: %v", err)
                 errResult = errors.Join(errResult, err)
                 continue
             }
             ...
         }
    }
    return errResult
```

Reply:
I agree that errors should be collected but this function should proceed for interfaces.